### PR TITLE
Runtime/Compiler - Support ComputeConfig::unpack_to_dest_mode #879

### LINKED ISSUE #879: Support `ComputeConfig::unpack_to_dest_mode`
Metal replaced `computeConfig.preserve_fp32_precision` with `unpack_to_dest_mode`:

```c++
enum class UnpackToDestMode : uint8_t    
{    
    UnpackToDestFp32,    
    Default    
}; 
```

```c++
struct ComputeConfig {
    ...
    std::vector<UnpackToDestMode> unpack_to_dest_mode;
};
```

We need to update:
- `TTKernelOpsEnums.td`: Add `UnpackToDestMode` enum
- `TTKernelOpsTypes.td`: Update `TensixConfigAttr` w/ new ArrayAttr field `unpack_to_dest_mode` (one per CB)
- `TTMetal/program.fbs`: Update `table TensixConfig`
- `ttmetal/command_queue.cpp`: Plumb through to metal api

### PR DESCRIPTION
**Pull request by kmabeeTT**
Note: This is blocked on tt-metal uplift. Will rebase + change target to main when tt-metal uplift is merged. For ease of viewing I picked base as my own branch for now on this PR.

Closes #879

Relatively simple change to accommodate API change in tt-metal introduced at https://github.com/tenstorrent/tt-metal/pull/12794

 - Pass field (per CB) down from compiler to runtime to metal API
 - Hardcoded to UnpackToDestMode::Default (preserve previous/default behavior) and for all 32 CBs for now because tt-metal requires this (will be relaxed soon), replace with compiler logic when ready.
 
 ==

About workaround:

1. Talked to Radomir who added constraint in tt-metal (field must be set for all 32 CBs) originally about relaxing it (  https://github.com/tenstorrent/tt-metal/issues/14066 )
2. Not sure what proper change in TTIRToTTMetal.cpp (to set field for only in-use CBs) will look like, it's a compiler task to figure out once above metal fix is pulled in.

### CODE REVIEW

**Review by AleksKnezevic — APPROVED** [review]

File: lib/Conversion/TTIRToTTMetal/TTIRToTTMetal.cpp [hunk 1]
```diff
@@ -1343,8 +1350,15 @@ class TTIRToTTMetalDispatchRewriter : public OpRewritePattern<ttir::GenericOp> {
     //   return failure();
     // }
 
+    // FIXME - Currently metal requires UnpackToDestMode for all 32 CBs but will
+    // be relaxed soon to remove this requirement. Update logic here later.
```
Comment (jnie-TT) [inline lib/Conversion/TTIRToTTMetal/TTIRToTTMetal.cpp:1354]: Worth adding this to the workarounds lib so we don't forget?

**Review by jnie-TT — APPROVED** [review]

File: lib/Conversion/TTIRToTTMetal/TTIRToTTMetal.cpp (hunk 1 above)
Comment (kmabeeTT) [reply lib/Conversion/TTIRToTTMetal/TTIRToTTMetal.cpp]: Removed fix/hack, found we have ~equivalent define in our side already, used it instead:

```
constexpr uint32_t kNumCBs = 1 + ttkernel::getMaxEnumValForCBPort();
```

Will add assert here to make sure our copy matches metal copy.

**Comment by kmabeeTT** [comment]
Cleaned up changes.  Had discussion with @nsmithtt offline this morning, decided to:

1. Don't need to hardcode using 32, instead we already have a tablegen-generated code to get max value of enum, use that instead
2. Add assert that our value of 32 (from say `1+ ttkernel::getMaxEnumValForCBPort()`) is equal to tt-metal's `NUM_CIRCULAR_BUFFERS=32`)
3. To do that, have tt-metal export that value (required in a tt-metal ticket already) and put it in our system-desc, and consume it in `TTIRToTTMetal.cpp`

Made changes for number-1 above already on PR, going to try to see what changes look like for 2,3 (with system-desc hardcoding num_circlular_buffers=32 until metal exports it).

File: include/ttmlir/Target/Common/types.fbs [hunk 2]
```diff
@@ -186,3 +187,8 @@ enum MathFidelity : uint8 {
     HiFi3         = 3,
     HiFi4         = 4,
 }
+
+enum UnpackToDestMode : uint8 {
+    UnpackToDestFp32 = 0,
+    Default = 1,
+}
```
Comment (nsmithtt) [inline include/ttmlir/Target/Common/types.fbs:194]: Should we move this to `program.fbs` since it's specific to programs?

**Review by nsmithtt — APPROVED** [review]

File: include/ttmlir/Target/Common/types.fbs (hunk 2 above)
Comment (kmabeeTT) [reply include/ttmlir/Target/Common/types.fbs]: Sure, good idea. Done (fixup, force push), this changed namespaces from:
```
tt::target::UnpackToDestMode::UnpackToDestFp32
```
to:

```
tt::target::metal::UnpackToDestMode::UnpackToDestFp32
```

**Review by tapspatel — APPROVED** [review]

**Pull request merged**